File: backend/security.py
from __future__ import annotations

import hashlib
import hmac
import secrets
import time
from collections import defaultdict, deque

from fastapi import HTTPException, Request, Response, status
from starlette.middleware.base import BaseHTTPMiddleware

from .config import get_settings

# ---------------------------------------------------------------------------
# Signed anonymous device id (used only to key non-sensitive quiz-history
# rows — which questions a browser has already seen — never for privileged
# access). Signed with HMAC so a client can't pick an arbitrary device id
# and tamper with someone else's rotation history.
# ---------------------------------------------------------------------------

DEVICE_COOKIE_NAME = "dynarctu_device"


def _sign(value: str, secret: str) -> str:
    mac = hmac.new(secret.encode(), value.encode(), hashlib.sha256).hexdigest()
    return f"{value}.{mac}"


def _verify(signed: str, secret: str) -> str | None:
    try:
        value, mac = signed.rsplit(".", 1)
    except ValueError:
        return None
    expected = hmac.new(secret.encode(), value.encode(), hashlib.sha256).hexdigest()
    if not hmac.compare_digest(mac, expected):
        return None
    return value


def get_or_create_device_id(request: Request, response: Response) -> str:
    settings = get_settings()
    raw = request.cookies.get(DEVICE_COOKIE_NAME)
    device_id = _verify(raw, settings.session_secret) if raw else None
    if not device_id:
        device_id = secrets.token_urlsafe(24)
        signed = _sign(device_id, settings.session_secret)
        response.set_cookie(
            key=DEVICE_COOKIE_NAME,
            value=signed,
            max_age=365 * 24 * 60 * 60,
            httponly=True,
            secure=settings.env != "development",
            samesite="lax",
            path="/",
        )
    return device_id


# ---------------------------------------------------------------------------
# Admin authentication (bearer token, constant-time compare).
# ---------------------------------------------------------------------------

def require_admin_token(request: Request) -> None:
    settings = get_settings()
    auth = request.headers.get("authorization", "")
    prefix = "Bearer "
    token = auth[len(prefix):] if auth.startswith(prefix) else ""
    if not token or not hmac.compare_digest(token, settings.admin_token):
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="unauthorized")


# ---------------------------------------------------------------------------
# Secure response headers.
# ---------------------------------------------------------------------------

class SecureHeadersMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next):
        settings = get_settings()
        response = await call_next(request)
        response.headers["X-Content-Type-Options"] = "nosniff"
        response.headers["X-Frame-Options"] = "DENY"
        response.headers["Referrer-Policy"] = "strict-origin-when-cross-origin"
        response.headers["Permissions-Policy"] = "geolocation=(), microphone=(), camera=(), payment=()"
        response.headers["Cross-Origin-Resource-Policy"] = "same-origin"
        response.headers["Content-Security-Policy"] = (
            "default-src 'self'; "
            "script-src 'self' https://cdn.jsdelivr.net; "
            "style-src 'self' 'unsafe-inline' https://fonts.googleapis.com; "
            "font-src https://fonts.gstatic.com https://cdn.jsdelivr.net; "
            "img-src 'self' data:; "
            "connect-src 'self'; "
            "object-src 'none'; "
            "base-uri 'none'; "
            "frame-ancestors 'none'"
        )
        if settings.enable_hsts:
            response.headers["Strict-Transport-Security"] = "max-age=63072000; includeSubDomains"
        if "server" in response.headers:
            del response.headers["server"]
        return response


# ---------------------------------------------------------------------------
# Lightweight in-process rate limiter (fixed window per client IP + bucket).
# Sufficient for the single-process "python server.py" deployment target;
# swap for a Redis-backed limiter if scaling out to multiple workers/hosts.
# ---------------------------------------------------------------------------

class RateLimiter:
    def __init__(self):
        self._hits: dict[str, deque] = defaultdict(deque)

    def check(self, bucket: str, client_key: str, limit_per_min: int) -> None:
        now = time.monotonic()
        window_start = now - 60
        key = f"{bucket}:{client_key}"
        hits = self._hits[key]
        while hits and hits[0] < window_start:
            hits.popleft()
        if len(hits) >= limit_per_min:
            raise HTTPException(
                status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                detail="too many requests, please slow down",
            )
        hits.append(now)


rate_limiter = RateLimiter()


def client_ip(request: Request) -> str:
    forwarded = request.headers.get("x-forwarded-for")
    if forwarded:
        return forwarded.split(",")[0].strip()
    return request.client.host if request.client else "unknown"

File: api/_lib/security.ts
import { createHmac, randomBytes, timingSafeEqual } from 'node:crypto';
import { parse as parseCookie, serialize as serializeCookie } from 'cookie';
import type { VercelRequest, VercelResponse } from '@vercel/node';
import { getSettings } from './config.js';

export const DEVICE_COOKIE_NAME = 'dynarctu_device';

function sign(value: string, secret: string): string {
  const mac = createHmac('sha256', secret).update(value).digest('hex');
  return `${value}.${mac}`;
}

function verify(signed: string, secret: string): string | null {
  const idx = signed.lastIndexOf('.');
  if (idx === -1) return null;
  const value = signed.slice(0, idx);
  const mac = signed.slice(idx + 1);
  const expected = createHmac('sha256', secret).update(value).digest('hex');
  const a = Buffer.from(mac);
  const b = Buffer.from(expected);
  if (a.length !== b.length || !timingSafeEqual(a, b)) return null;
  return value;
}

/**
 * Reads (or mints) a signed, anonymous per-browser device id, used only to
 * key non-sensitive quiz-history rows (which questions a browser has
 * already seen) — never for privileged access. Signed with HMAC so a
 * client can't pick an arbitrary device id and tamper with someone else's
 * rotation history.
 */
export function getOrCreateDeviceId(req: VercelRequest, res: VercelResponse): string {
  const settings = getSettings();
  const cookies = req.cookies || parseCookie(req.headers.cookie || '');
  const raw = cookies[DEVICE_COOKIE_NAME];
  let deviceId = raw ? verify(raw, settings.sessionSecret) : null;

  if (!deviceId) {
    deviceId = randomBytes(24).toString('base64url');
    const signed = sign(deviceId, settings.sessionSecret);
    const cookieStr = serializeCookie(DEVICE_COOKIE_NAME, signed, {
      maxAge: 365 * 24 * 60 * 60,
      httpOnly: true,
      secure: settings.env !== 'development',
      sameSite: 'lax',
      path: '/',
    });
    appendSetCookie(res, cookieStr);
  }
  return deviceId;
}

function appendSetCookie(res: VercelResponse, cookieStr: string): void {
  const existing = res.getHeader('Set-Cookie');
  if (!existing) {
    res.setHeader('Set-Cookie', cookieStr);
  } else if (Array.isArray(existing)) {
    res.setHeader('Set-Cookie', [...existing, cookieStr]);
  } else {
    res.setHeader('Set-Cookie', [existing as string, cookieStr]);
  }
}

/** Constant-time bearer-token check for /api/admin/* routes. */
export function requireAdminToken(req: VercelRequest): void {
  const settings = getSettings();
  const auth = (req.headers.authorization as string) || '';
  const prefix = 'Bearer ';
  const token = auth.startsWith(prefix) ? auth.slice(prefix.length) : '';
  const a = Buffer.from(token);
  const b = Buffer.from(settings.adminToken);
  if (!token || a.length !== b.length || !timingSafeEqual(a, b)) {
    const err = new Error('unauthorized') as Error & { statusCode: number };
    err.statusCode = 401;
    throw err;
  }
}

export function clientIp(req: VercelRequest): string {
  const forwarded = req.headers['x-forwarded-for'];
  if (forwarded) {
    const value = Array.isArray(forwarded) ? forwarded[0] : forwarded;
    return value.split(',')[0].trim();
  }
  return req.socket?.remoteAddress || 'unknown';
}

/** Same header set as the original SecureHeadersMiddleware. */
export function applySecurityHeaders(req: VercelRequest, res: VercelResponse): void {
  const settings = getSettings();
  res.setHeader('X-Content-Type-Options', 'nosniff');
  res.setHeader('X-Frame-Options', 'DENY');
  res.setHeader('Referrer-Policy', 'strict-origin-when-cross-origin');
  res.setHeader('Permissions-Policy', 'geolocation=(), microphone=(), camera=(), payment=()');
  res.setHeader('Cross-Origin-Resource-Policy', 'same-origin');
  res.setHeader(
    'Content-Security-Policy',
    "default-src 'self'; " +
      "script-src 'self' https://cdn.jsdelivr.net; " +
      "style-src 'self' 'unsafe-inline' https://fonts.googleapis.com; " +
      'font-src https://fonts.gstatic.com https://cdn.jsdelivr.net; ' +
      "img-src 'self' data:; " +
      "connect-src 'self'; " +
      "object-src 'none'; " +
      "base-uri 'none'; " +
      "frame-ancestors 'none'"
  );
  if (settings.enableHsts) {
    res.setHeader('Strict-Transport-Security', 'max-age=63072000; includeSubDomains');
  }
  if (settings.allowedOrigins.length) {
    const origin = (req.headers.origin as string) || '';
    if (settings.allowedOrigins.includes(origin)) {
      res.setHeader('Access-Control-Allow-Origin', origin);
      res.setHeader('Access-Control-Allow-Credentials', 'true');
      res.setHeader('Vary', 'Origin');
    }
  }
  // Same-origin default (no ALLOWED_ORIGINS set): no CORS headers are added
  // at all, so the browser's same-origin policy blocks all cross-origin
  // access to the API — correct default for an app that ships its own
  // frontend from the same Vercel deployment/domain.
}
